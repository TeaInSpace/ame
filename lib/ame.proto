syntax = "proto3";

package ame.v1 ;

message TaskIdentifier {
  string name = 1;
}

enum TaskType {
  Pipenv = 0;
  Mlflow = 1;
  Poetry = 2;
}

message Empty {
}

message ProjectFileIdentifier {
  string taskid= 1;
  string filepath = 2;
}

message FileChunk {
  bytes contents = 1;
}

message TaskProjectDirectoryStructure {
  string projectid = 1;
  TaskIdentifier taskid = 2; 
  repeated string paths = 3;
}

message ProjectFileChunk {
  oneof messages {
    FileChunk chunk = 1;
    ProjectFileIdentifier identifier = 2;
  }
}

message TaskLogRequest {
  TaskIdentifier taskid = 1;
  optional int32 start_from = 2;
  optional bool watch = 3;
}

message LogEntry {
  bytes contents = 1;
}

message ProjectSourceCfg {
  optional GitProjectSource git = 1;
}

message GitProjectSource {
  string repository = 1;
  optional string username = 2;
  optional string secret = 3;
  optional string sync_interval = 4;
}

enum ProjectSourceState {
  Pending = 0;
  Synchronising = 1;
  Synchronized = 2;
  Error = 3;
}

enum ProjectSourceIssueType {
  Unknown = 0;
  AuthFailure = 1;
  RepositoryNotFound = 2;
  AmeProjectNotFound = 3;
  GitSecretNotFound = 4;
}

message ProjectSourceStatus {
  optional string last_synced = 1;
  ProjectSourceState state = 2;
  optional string reason = 3;
  repeated ProjectSourceIssue issues = 4;
}

message ProjectSourceIssue {
  ProjectSourceIssueType issue_type = 1;
  optional string explanation = 2;
}

message ProjectSourceId {
  string name = 1;
}

message TrainRequest {
  string projectid = 1;
  string model_name = 2;
}

// This is a secret stored by AME
message AmeSecret {
  string key = 1;
  string value = 2;
}

message AmeSecretId {
  string key = 1;
}

message AmeSecrets {
repeated AmeSecretId secrets = 1;
}

message ResourceCfg {
  oneof cfg {
    ProjectSourceCfg projectSrcCfg = 1;
  }
}

message ResourceId {
  oneof id {
    ProjectSourceId projectSrcId = 1;
  }
}


message ProjectSourceListParams {}

message ResourceListParams {
  oneof params {
    ProjectSourceListParams projectSourceListParams = 1;
  }
}

message ResourceIds {
  repeated ResourceId ids = 1;
}

message ListProjectSrcsResponse {
  repeated ProjectSourceCfg cfgs = 2;
}

message ProjectSrcPatchRequest {
  ProjectSourceId id = 1;
  ProjectSourceCfg cfg = 2;
}

message ProjectSrcIdRequest {
  string repo = 1;
}

message TaskRef {
  string name = 1;
  optional string project = 2;
}

message TaskCfg {
  optional string name = 1;
  optional TaskRef taskRef = 2;  
  map<string, string> resources = 3;
  oneof executor {
    PoetryExecutor poetry = 4; 
    MlflowExecutor mlflow = 5;
    PipEnvExecutor pipEnv = 6;
    PipExecutor pip = 8;
    CustomExecutor custom = 9;
  }
  repeated string dataSets = 7;
  optional TemplateRef fromTemplate = 10;
  optional ArtifactCfg artifactCfg = 11;
  optional TriggerCfg triggers = 12;
  repeated EnvVar env = 13;
  repeated Secret secrets = 14;
}

message Secret {
  oneof variant {
    AmeSecretVariant ame = 1;
  }
}

message AmeSecretVariant {
  string key = 1;
  string injectAs = 2;
}

message EnvVar {
  string key = 1;
  string val = 2;
}

message TriggerCfg {
  optional string schedule = 1;
}

message ArtifactCfg {
  bool save_changed_files = 1;
  repeated string paths = 2; 
}

message TemplateRef {
  string name = 1;
  optional string project = 2;
}

message CustomExecutor {
  string pythonVersion = 1;
  string command = 2;
}

message PoetryExecutor {
  string pythonVersion = 1;  
  string command = 2;
}


message PipExecutor {
  string pythonVersion = 1;  
  string command = 2;
}

message MlflowExecutor {
  
}

message PipEnvExecutor {
  string command = 1;
}

// TODO: should there be an error case?
message TaskStatus {
  oneof phase {
    TaskPhasePending pending = 2;  
    TaskPhaseRunning running = 3;
    TaskPhaseFailed failed = 4;
    TaskPhaseSucceeded succeeded= 5;
  }
}

message TaskPhasePending {
  
}

message TaskPhaseRunning {
  string workflowName= 1;
}
message TaskPhaseFailed {
  string workflowName= 1;
}
message TaskPhaseSucceeded {
  string workflowName= 1;
}

message DataSetCfg {
 string name = 1;
 TaskCfg  task = 2;
 string path = 3;
 optional string size = 4;
}

message ProjectCfg {
  string name = 1;
  repeated Model models = 2;
  repeated DataSetCfg dataSets = 3;
  repeated TaskCfg tasks = 4;
  repeated TaskCfg templates = 5;
  optional bool enableTriggers = 6;
}

message ProjectStatus {
  map<string,ModelStatus> models = 1;
}

message ModelStatus {
  optional string latestValidatedModelVersion = 1;
}

message CreateProjectRequest {
  ProjectCfg cfg = 1;
  optional bool enableTriggers = 2;
}

message ProjectId {
  string name = 1;
}

message Model {
  string name = 1;
  optional TaskCfg validationTask = 2; 
  optional ModelTrainingCfg training = 3;
  optional ModelDeploymentCfg deployment = 4;
}

message ModelTrainingCfg {
  TaskCfg task = 1;
}

message ModelDeploymentCfg {
   map<string, string> ingressAnnotations = 1;
   optional int32 replicas = 2;
   optional string image = 3;
   map<string,string> resources = 4;
   optional bool enableTls = 5;
}

message RunTaskRequest {
  ProjectId projectId= 1;
  TaskCfg taskCfg = 2;
}

message TaskId {
  string name = 1;
}

message ListTasksRequest {
  
}

message ListTasksResponse {
  map<string, TaskListEntry> tasks = 1;
}

message TaskListEntry {
  TaskStatus status = 1;
  string timeStamp = 2;
}

message RemoveTaskRequest {
  string name = 1;
  optional bool approve = 2;
}

service AmeService {
  rpc RunTask(RunTaskRequest) returns (TaskIdentifier) {}
  rpc GetTask(TaskIdentifier) returns (TaskCfg) {}
  rpc DeleteTask(TaskIdentifier) returns (Empty) {}
  rpc CreateTaskProjectDirectory(TaskProjectDirectoryStructure) returns (Empty) {}
  rpc UploadProjectFile(stream ProjectFileChunk) returns (Empty) {}
  rpc StreamTaskLogs(TaskLogRequest) returns (stream LogEntry) {}
  rpc CreateProjectSrc(ProjectSourceCfg) returns (ProjectSourceId) {}
  rpc DeleteProjectSrc(ProjectSourceId) returns (Empty) {}
  rpc TrainModel(TrainRequest) returns (Empty) {}
  rpc WatchProjectSrc(ProjectSourceId) returns (stream ProjectSourceStatus) {}
  rpc CreateSecret(AmeSecret) returns (Empty) {}
  rpc DeleteSecret(AmeSecretId) returns (Empty) {}
  rpc ListSecrets(Empty) returns (AmeSecrets) {} 
  rpc UpdateProjectSrc(ProjectSrcPatchRequest) returns (Empty) {}
  rpc CreateResource(ResourceCfg) returns (ResourceId) {}
  rpc ListResource(ResourceListParams) returns (ResourceIds) {}
  rpc GetProjectSrcCfg(ProjectSourceId) returns (ProjectSourceCfg) {}
  rpc GetProjectSrcStatus(ProjectSourceId) returns (ProjectSourceStatus) {}
  rpc GetProjectSrcId(ProjectSrcIdRequest) returns (ProjectSourceId) {}
  rpc ListProjectSrcs(ProjectSourceListParams) returns (ListProjectSrcsResponse) {} 
  rpc CreateProject(CreateProjectRequest) returns (ProjectId) {}
  rpc ListTasks(ListTasksRequest) returns (ListTasksResponse) {}
  rpc RemoveTask(RemoveTaskRequest) returns (Empty) {}
 }
