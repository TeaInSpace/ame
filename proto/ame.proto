syntax = "proto3";

package ame.v1 ;

message TaskIdentifier {
  string name = 1;
}

enum TaskType {
  Pipenv = 0;
  Mlflow = 1;
}

message TaskTemplate {
  string projectid = 1;
  string command = 2;
  string name = 3;
  optional string image = 4;
  optional TaskType taskType = 5;
}

message CreateTaskRequest {
  TaskIdentifier id = 1;
  TaskTemplate template = 2;
}

message Empty {
}

message ProjectFileIdentifier {
  string taskid= 1;
  string filepath = 2;
}

message FileChunk {
  bytes contents = 1;
}

message TaskProjectDirectoryStructure {
  string projectid = 1;
  TaskIdentifier taskid = 2; 
  repeated string paths = 3;
}

message ProjectFileChunk {
  oneof messages {
    FileChunk chunk = 1;
    ProjectFileIdentifier identifier = 2;
  }
}

message TaskLogRequest {
  TaskIdentifier taskid = 1;
  optional int32 start_from = 2;
  optional bool watch = 3;
}

message LogEntry {
  bytes contents = 1;
}

message ProjectSource {
  optional GitProjectSource git = 1;
}

message GitProjectSource {
  string repository = 1;
}

service AmeService {
  rpc GetTask(TaskIdentifier) returns (TaskTemplate) {}
  rpc CreateTask(CreateTaskRequest) returns (TaskIdentifier) {}
  rpc DeleteTask(TaskIdentifier) returns (Empty) {}
  rpc CreateTaskProjectDirectory(TaskProjectDirectoryStructure) returns (Empty) {}
  rpc UploadProjectFile(stream ProjectFileChunk) returns (Empty) {}
  rpc StreamTaskLogs(TaskLogRequest) returns (stream LogEntry) {}
}
