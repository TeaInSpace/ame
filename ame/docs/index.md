# AME

## What is AME?

AME(Artificial MLOps Engineer) automates operations for the entire model life cycle from early experimentation to deploying models for inferece.

AME abstracts away details regarding infrastructure allowing data scientists to focus on datascience.

Getting a project started with AME can be as simple as:

```sh
ame init myproject
ame add task training
ame task run training
```

## How does AME differ from existing MLOps tools?


