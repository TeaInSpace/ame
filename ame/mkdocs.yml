site_name: AME 
theme: 
  name: material
  features:
    - navigation.tabs
  palette:
     # Palette toggle for light mode
    - scheme: default
      primary: green 
      accent: light green 
      toggle:
        icon: material/brightness-7 

        name: Switch to dark mode

    # Palette toggle for dark mode
    - scheme: slate
      primary: green 
      accent: light green
      toggle:
        icon: material/brightness-4
        name: Switch to light mode

markdown_extensions:
  - pymdownx.highlight:
      anchor_linenums: true
      line_spans: __span
      pygments_lang_class: true
  - pymdownx.inlinehilite
  - pymdownx.snippets
  - pymdownx.superfences

