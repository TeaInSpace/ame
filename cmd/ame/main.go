package main

import (
	commands "teainspace.com/ame/cmd/ame/cmd"
)

func main() {
	commands.Execute()
}

func run() int {
	commands.Execute()
	return 0
}
