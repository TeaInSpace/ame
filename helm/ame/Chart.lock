dependencies:
- name: ingress-nginx
  repository: https://kubernetes.github.io/ingress-nginx
  version: 4.7.1
- name: minio
  repository: oci://registry-1.docker.io/bitnamicharts
  version: 12.6.12
- name: argo-workflows
  repository: https://argoproj.github.io/argo-helm
  version: 0.32.1
- name: mlflow
  repository: https://community-charts.github.io/helm-charts
  version: 0.7.19
digest: sha256:643139c7e6ed8564f598cb8513443449755f59b8ae016c8171d3be94bdf0018b
generated: "2023-08-08T14:50:26.129967853+02:00"
