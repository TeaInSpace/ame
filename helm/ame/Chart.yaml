apiVersion: v2
name: ame
description: A helm chart AME the artificial MLOps engineer.

# A chart can be either an 'application' or a 'library' chart.
#
# Application charts are a collection of templates that can be packaged into versioned archives
# to be deployed.
#
# Library charts provide useful utilities or functions for the chart developer. They're included as
# a dependency of application charts to inject those utilities and functions into the rendering
# pipeline. Library charts do not define any templates and therefore cannot be deployed.
type: application

# This is the chart version. This version number should be incremented each time you make changes
# to the chart and its templates, including the app version.
# Versions are expected to follow Semantic Versioning (https://semver.org/)
version: 0.1.0-alpha1

# This is the version number of the application being deployed. This version number should be
# incremented each time you make changes to the application. Versions are not expected to
# follow Semantic Versioning. They should reflect the version the application is using.
# It is recommended to use it with quotes.
appVersion: "0.1.0-alpha1"

dependencies:
- name: ingress-nginx
  version: "4.7.*"
  repository: "https://kubernetes.github.io/ingress-nginx"
  condition: nginxIngress.enabled
- name: minio
  repository: "oci://registry-1.docker.io/bitnamicharts"
  version: "12.6.12"
  condition: minio.enabled
- name: argo-workflows
  repository: "https://argoproj.github.io/argo-helm"
  version: "0.32.1"
  condition: argoWorkflows.enabled
- name: mlflow
  repository: "https://community-charts.github.io/helm-charts"
  version: 0.7.19
  condition: mlflow.enabled
