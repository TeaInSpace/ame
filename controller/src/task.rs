use std::{
    sync::Arc,
    time::{self, Duration},
};

use ame::{
    ctrl::AmeResource,
    custom_resources::{
        argo::{Workflow, WorkflowPhase},
        data_set::{DataSet, DataSetPhase, DataSetStatus},
        find_project,
        new_task::{build_workflow, resolve_task_templates, Task},
        project::{local_name, project_name, Project},
        task_ctrl::TaskCtrl,
    },
    error::AmeError,
    grpc::{task_status::Phase, TaskPhaseFailed, TaskPhaseRunning, TaskPhaseSucceeded, TaskStatus},
    Result,
};
use envconfig::Envconfig;
use futures::{future::BoxFuture, FutureExt, StreamExt};
use kube::{
    api::{ListParams, Patch, PatchParams},
    runtime::{controller::Action, finalizer, Controller},
    Api, Client, ResourceExt,
};
use tracing::{debug, error, info};

static TASK_CONTROLLER: &str = "tasks.ame.teainspace.com";

#[derive(Clone)]
struct Context {
    cfg: TaskControllerCfg,
    client: Client,
}

impl Context {
    fn new(client: Client, cfg: TaskControllerCfg) -> Self {
        Self { client, cfg }
    }
}

#[derive(Clone, Envconfig, Debug)]
pub struct TaskControllerCfg {
    #[envconfig(
        from = "AME_EXECUTOR_IMAGE",
        default = "main.localhost:45373/ame-executor:latest"
    )]
    pub executor_image: String,
    #[envconfig(from = "NAMESPACE")]
    pub namespace: Option<String>,
    #[envconfig(from = "AME_SERVICE_ACCOUNT", default = "ame-task")]
    pub service_account: String,
}

async fn reconcile(task: Arc<Task>, ctx: Arc<Context>) -> Result<Action> {
    info!("reconciling task {}", task.name_any());

    let tasks = if let Some(ref namespace) = ctx.cfg.namespace {
        Api::<Task>::namespaced(ctx.client.clone(), namespace)
    } else {
        todo!("we need to handle this case better??");
    };

    Ok(finalizer(&tasks, TASK_CONTROLLER, task, |event| async {
        match event {
            finalizer::Event::Apply(task) => apply(&task, &ctx).await,
            finalizer::Event::Cleanup(task) => cleanup(&task, &tasks).await,
        }
    })
    .await?)
}

// TODO: do not allow nonexistent fields in project.yaml.

async fn apply(task: &Task, ctx: &Context) -> Result<Action> {
    let (tasks, workflows, projects, data_sets) = if let Some(ref namespace) = ctx.cfg.namespace {
        (
            Api::<Task>::namespaced(ctx.client.clone(), namespace),
            Api::<Workflow>::namespaced(ctx.client.clone(), namespace),
            Api::<Project>::namespaced(ctx.client.clone(), namespace),
            Api::<DataSet>::namespaced(ctx.client.clone(), namespace),
        )
    } else {
        todo!("we need to handle this case better??");
    };

    let task_ctrl = TaskCtrl::new(data_sets.clone(), projects.clone());

    debug!("checking datasets for task {:?}", task.name_any());

    let project = projects.get(&task.parent_project_name()?).await?;

    if !task.spec.cfg.data_sets.is_empty() {
        info!(
            "reconciling data sets for task {}",
            task.spec.cfg.name.as_ref().unwrap()
        );

        let mut ds_statuses: Vec<Option<DataSetStatus>> = vec![];

        debug!("reconciling datasets {:?}", task.spec.cfg.data_sets);

        for ds in task.spec.cfg.data_sets.iter() {
            let ds_name = local_name(ds.clone());

            let ds_project = if let Some(project_name) = project_name(ds.clone()) {
                find_project(projects.clone(), project_name, "".to_string())
                    .await
                    .or(Err(AmeError::MissingProject(0)))?
            } else {
                project.clone()
            };

            let mut data_set = match ds_project.generate_data_set(ds_name.clone()) {
                Ok(ds) => ds,
                Err(e) => {
                    error!(
                        "failed to reconcile data set {} for project {}: {}",
                        ds_name,
                        ds_project.name_any(),
                        e
                    );
                    continue;
                }
            };
            let Some(task_oref) = task.gen_owner_ref() else {
                error!(
                    "failed to generate owner reference for task {} stopping data set creation",
                    task.name_any()
                );
                continue;
            };

            data_set.owner_references_mut().push(task_oref);

            debug!("patching dataset: {}", ds);

            let ds_obj_name =
                data_set
                    .metadata
                    .name
                    .clone()
                    .ok_or(AmeError::ReconcilitationFailure(
                        "Task".to_string(),
                        task.name_any(),
                        format!(
                            "data set  {} in project {} is missing a name",
                            ds, project.spec.cfg.name
                        ),
                    ))?;

            let data_set = data_sets
                .patch(
                    &ds_obj_name,
                    &PatchParams::apply(TASK_CONTROLLER).force(),
                    &Patch::Apply(data_set),
                )
                .await?;

            ds_statuses.push(data_set.status);
        }

        for stat in ds_statuses {
            let Some(stat) = stat else {
                info!("waiting for datasets to complete");
                return Ok(Action::requeue(Duration::from_secs(10)));
            };

            match stat.phase {
                Some(DataSetPhase::Ready { .. }) => continue,
                Some(DataSetPhase::Failed { .. }) => {
                    error!(
                        "Data set has failed, can not schedule Task {}",
                        task.spec
                            .cfg
                            .name
                            .as_ref()
                            .unwrap_or(&"unknown name".to_string())
                    );
                }
                _ => {
                    info!("waiting for datasets to complete");
                    return Ok(Action::requeue(Duration::from_secs(10)));
                }
            }
        }
    }

    let task_ctx = task_ctrl
        .gather_task_ctx(
            task,
            ctx.cfg.executor_image.to_string(),
            ctx.cfg.service_account.clone(),
        )
        .await?;

    debug!("resolving task {}", task.name_any());

    let resolved_task = resolve_task_templates(task.clone(), project, projects).await?;

    debug!("resolved task {:?}", task.spec.cfg);

    let workflow = build_workflow(resolved_task, task_ctx)?;

    debug!("patching workflow for task {:?} ", task.name_any(),);

    let workflow = workflows
        .patch(
            &workflow.name_any(),
            &PatchParams::apply(TASK_CONTROLLER).force(),
            &Patch::Apply(&workflow),
        )
        .await?;

    debug!("workflow phase: {:?}", workflow.status);

    let phase = workflow
        .status
        .as_ref()
        .map(|s| match s.phase {
            WorkflowPhase::Pending | WorkflowPhase::Running | WorkflowPhase::Error => {
                Phase::Running(TaskPhaseRunning {
                    workflow_name: workflow.name_any(),
                })
            }
            WorkflowPhase::Failed => Phase::Failed(TaskPhaseFailed {
                workflow_name: workflow.name_any(),
            }),
            WorkflowPhase::Succeeded => Phase::Succeeded(TaskPhaseSucceeded {
                workflow_name: workflow.name_any(),
            }),
        })
        .unwrap_or(Phase::Running(TaskPhaseRunning {
            workflow_name: workflow.name_any(),
        }));

    let mut task = task.clone();

    task.status = Some(TaskStatus { phase: Some(phase) });
    task.metadata.managed_fields = None;

    debug!("patching status for task {}", task.name_any());
    tasks
        .patch_status(
            &task.name_any(),
            &PatchParams::apply(TASK_CONTROLLER).force(),
            &Patch::Apply(task),
        )
        .await?;

    Ok(Action::requeue(std::time::Duration::from_secs(60)))
}

pub async fn cleanup(task: &Task, _tasks: &Api<Task>) -> Result<Action> {
    info!("cleanup dataset: {}", task.name_any());

    if task.spec.deletion_approved {
        Ok(Action::requeue(time::Duration::from_secs(300)))
    } else {
        Err(AmeError::ApiError("deletion not approved".to_string()))
    }
}

pub async fn start_task_controller(
    client: Client,
    config: TaskControllerCfg,
) -> Result<BoxFuture<'static, ()>> {
    info!("Start Task controller");
    let context = Arc::new(Context::new(client.clone(), config.clone()));

    let tasks = if let Some(ref namespace) = config.namespace {
        Api::<Task>::namespaced(client.clone(), namespace)
    } else {
        Api::<Task>::all(client.clone())
    };

    let workflow = if let Some(ref namespace) = config.namespace {
        Api::<Workflow>::namespaced(client, namespace)
    } else {
        Api::<Workflow>::all(client)
    };

    Ok(Controller::new(tasks, ListParams::default())
        .owns(workflow, ListParams::default())
        .run(reconcile, error_policy, context)
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .boxed())
}

fn error_policy(_src: Arc<Task>, error: &AmeError, _ctx: Arc<Context>) -> Action {
    error!("failed to reconcile: {:?}", error);
    Action::requeue(Duration::from_secs(5 * 60))
}

#[cfg(test)]
mod test {
    use std::collections::BTreeMap;

    use ame::{
        custom_resources::{
            argo::Workflow,
            new_task::{ProjectSource, TaskSpec},
        },
        grpc::{
            task_cfg::Executor, task_status::Phase, PoetryExecutor, ProjectCfg, TaskCfg,
            TaskPhaseRunning,
        },
        Result,
    };
    use envconfig::Envconfig;

    use kube::{
        api::{DeleteParams, PostParams},
        core::ObjectMeta,
        Api, Client, Resource,
    };

    use super::*;

    #[tokio::test]
    #[ignore = "requires a k8s cluster"]
    async fn can_create_workflow_and_finalize_task() -> Result<()> {
        let client = Client::try_default().await?;
        let namespace = "default".to_string();
        let tasks = Api::<Task>::namespaced(client.clone(), &namespace);
        let workflows = Api::<Workflow>::namespaced(client.clone(), &namespace);
        let projects = Api::<Project>::namespaced(client.clone(), &namespace);

        let mut controller_cfg = TaskControllerCfg::init_from_env().unwrap();
        controller_cfg.namespace = Some(namespace);

        let ctx = Context::new(client, controller_cfg);

        let project = ProjectCfg {
            name: "parentproject".to_string(),
            models: vec![],
            data_sets: vec![],
            tasks: vec![],
            templates: vec![],
            enable_triggers: None,
        };

        let project = Project::from_cfg(project);
        let project = projects.create(&PostParams::default(), &project).await?;

        let task = Task {
            metadata: ObjectMeta {
                generate_name: Some("mytask".to_string()),
                owner_references: Some(vec![project.controller_owner_ref(&()).unwrap()]),
                ..ObjectMeta::default()
            },
            spec: TaskSpec {
                cfg: TaskCfg {
                    name: Some("mytask".to_string()),
                    task_ref: None,
                    executor: Some(Executor::Poetry(PoetryExecutor {
                        python_version: "3.11".to_string(),
                        command: "python train.py".to_string(),
                    })),
                    resources: BTreeMap::new(),
                    data_sets: Vec::new(),
                    from_template: None,
                    artifact_cfg: None,
                    triggers: None,
                    env: vec![],
                    secrets: vec![],
                },
                source: Some(ProjectSource::Ame {
                    path: "somepath".to_string(),
                }),
                deletion_approved: false,
                project: None,
            },
            status: None,
        };

        let task = tasks.create(&PostParams::default(), &task).await?;

        reconcile(Arc::new(task.clone()), Arc::new(ctx.clone())).await?;

        let task = tasks.get(&task.name_any()).await?;

        reconcile(Arc::new(task.clone()), Arc::new(ctx.clone())).await?;

        let task_status = tasks.get_status(&task.name_any()).await?.status.unwrap();

        let Phase::Running(TaskPhaseRunning { workflow_name }) = task_status.phase.unwrap() else {
            panic!("task was not running");
        };

        let workflow = workflows.get(&workflow_name).await?;

        let mut settings = insta::Settings::clone_current();

        settings.add_filter(
            &format!("{}.+", task.clone().metadata.generate_name.unwrap()),
            "redacted",
        );

        let _guard = settings.bind_to_scope();
        insta::assert_yaml_snapshot!(&workflow.spec);

        tasks
            .delete(&task.name_any(), &DeleteParams::default())
            .await?;

        let data_set = tasks.get(&task.name_any()).await?;

        // An error is expected here as the cleanup is rejected.
        reconcile(Arc::new(data_set.clone()), Arc::new(ctx.clone()))
            .await
            .unwrap_err();

        // Deletion should not be possible until deletion has been approved.
        let mut task = tasks.get(&task.name_any()).await?;

        // Approve deletion and verify that the data set is now deleted.
        task.spec.deletion_approved = true;

        tasks
            .patch(
                &task.name_any(),
                &PatchParams::default(),
                &Patch::Merge(task.clone()),
            )
            .await?;

        reconcile(Arc::new(task.clone()), Arc::new(ctx.clone()))
            .await
            .unwrap();

        // The cluster is allowed a minimum of 100ms to delete the data set
        // object after the finalizer has approved deletion during the last
        // reconcile call.
        tokio::time::sleep(Duration::from_millis(100)).await;

        tasks.get(&task.name_any()).await.unwrap_err();

        Ok(())
    }
}
