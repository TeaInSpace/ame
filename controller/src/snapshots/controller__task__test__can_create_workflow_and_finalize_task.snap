---
source: controller/src/task.rs
expression: "&workflow.spec"
---
entrypoint: main
templates:
  - name: main
    metadata:
      labels: ~
      annotations: ~
    steps:
      - - name: setup
          inline:
            name: setup
            metadata:
              labels:
                ame-task: redacted
              annotations: ~
            steps: ~
            securityContext:
              fsGroup: 2000
              runAsUser: 1001
            script:
              command:
                - bash
              env:
                - name: AWS_ACCESS_KEY_ID
                  valueFrom:
                    secretKeyRef:
                      key: MINIO_ROOT_USER
                      name: ame-minio-secret
                      optional: false
                - name: AWS_SECRET_ACCESS_KEY
                  valueFrom:
                    secretKeyRef:
                      key: MINIO_ROOT_PASSWORD
                      name: ame-minio-secret
                      optional: false
                - name: MLFLOW_TRACKING_URI
                  value: "http://mlflow.default.svc.cluster.local:5000"
                - name: MINIO_URL
                  value: "http://ame-minio.ame-system.svc.cluster.local:9000"
                - name: PIPENV_YES
                  value: "1"
              image: "main.localhost:45373/ame-executor:latest"
              name: ""
              resources:
                limits: {}
              volumeMounts:
                - mountPath: /project
                  name: redacted
              source: " \n\n s3cmd --no-ssl --region eu-central-1 --host=$MINIO_URL --host-bucket=$MINIO_URL get --recursive s3://somepath ./"
            podSpecPatch: ~
      - - name: redacted
          inline:
            name: redacted
            metadata:
              labels:
                ame-task: redacted
              annotations: ~
            steps: ~
            securityContext:
              fsGroup: 2000
              runAsUser: 1001
            script:
              command:
                - bash
              env:
                - name: AWS_ACCESS_KEY_ID
                  valueFrom:
                    secretKeyRef:
                      key: MINIO_ROOT_USER
                      name: ame-minio-secret
                      optional: false
                - name: AWS_SECRET_ACCESS_KEY
                  valueFrom:
                    secretKeyRef:
                      key: MINIO_ROOT_PASSWORD
                      name: ame-minio-secret
                      optional: false
                - name: MLFLOW_TRACKING_URI
                  value: "http://mlflow.default.svc.cluster.local:5000"
                - name: MINIO_URL
                  value: "http://ame-minio.ame-system.svc.cluster.local:9000"
                - name: PIPENV_YES
                  value: "1"
              image: "main.localhost:45373/ame-executor:latest"
              name: ""
              resources:
                limits: {}
              volumeMounts:
                - mountPath: /project
                  name: redacted
              source: "\n                        source ~/.bashrc\n                        \n                        pyenv install 3.11\n\n                        pyenv global 3.11\n\n                        poetry install\n                    \n                        poetry run python train.py\n                    "
            podSpecPatch: ~
    securityContext: ~
    script: ~
    podSpecPatch: ~
imagePullSecrets: ~
volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: redacted
    spec:
      accessModes:
        - ReadWriteOnce
      resources:
        requests:
          storage: 50Gi
    status: {}
volumes: ~
serviceAccountName: ame-task

